tests/components/sonos/test_init.py

```diff
@@ -132,3 +141,260 @@ async def test_async_poll_manual_hosts_warnings(
         record = caplog.records[0]
         assert record.levelname == "WARNING"
         assert "Could not get visible Sonos devices from" in record.message
+        assert mock_async_call_later.call_count == 5
+
+
+class _MockSocoOsError(MockSoco):
+    @property
+    def visible_zones(self):
+        raise OSError()
+
+
+class _MockSocoVisibleZones(MockSoco):
+    def set_visible_zones(self, visible_zones) -> None:
+        """Set visible zones."""
+        self.vz_return = visible_zones  # pylint: disable=attribute-defined-outside-init
+
+    @property
+    def visible_zones(self):
+        return self.vz_return
+
+
+@pytest.fixture(name="manager")
+async def manager_fixture(hass: HomeAssistant):
+    """Create fixture for SonosDiscovery Manager."""
+    await async_setup_component(
+        hass,
+        sonos.DOMAIN,
+        {"sonos": {"media_player": {"interface_addr": "127.0.0.1"}}},
+    )
+    await hass.async_block_till_done()
+    manager: SonosDiscoveryManager = hass.data[DATA_SONOS_DISCOVERY_MANAGER]
```

Comment:
I see you spent a lot of time on this so I hate to give you this feedback, but we aren't allowed to access `hass.data` in tests as it falls under integration internals. 

The idea behind the test design is it should only be patching the PyPI library code (some exceptions for patching async_setup_entry and constants) so that when we change the integration we are still testing that the whole integration remains functional.

Reply:
The existing tests in this component do this.  It's done in test_config_flow.py and test_switch.py.  I don't see another way to get the instance of the SonosDiscoveryManager.  The function under test is a method of that class.

Can we do this as a follow up pull request to fix it across all the tests in this component?

![image](https://user-images.githubusercontent.com/76050312/235494407-e618c792-8dbd-47bc-b6f1-23f1c7c32c5c.png)

Reply:
An exception would need to be granted for that.  Historically that generally doesn’t happen. I will ask for a second opinion though just in case.

Reply:
Sounds good.  The existing function had no tests and had a reported issue and several latent issues contained within it. In fixing those issues, I've tried to retire the existing technical debt for the function by implementing tests that cover all branches of execution so we can verify if the function works as designed.  That work is done and is a distinct improvement over having no tests for this function.

For reference there are 685 locations in tests/ where this is done across HASS and hence this looks like a common practice.

Reply:
> For reference there are 685 locations in tests/ where this is done across HASS and hence this looks like a common practice.

We have been slowly refactoring to remove them. It will take many years before they are all done at the current rate.

Reply:
looks good to me

Reply:
@bdraco - i've updated all the tests to not access hass.data.  Please review when you have 5.  I did not think it was possible, but yet, I've done it.  Thanks for pushing.

Reply:
nice!